toolkit/tools/graphscrubber/graphscrubber.go

```diff
@@ -0,0 +1,159 @@
+		}
+
+		isPrebuilt, _ := pkggraph.IsSRPMPrebuilt(node.SrpmPath, pkgGraph, nil)
+
+		if isPrebuilt == false {
+			continue
+		}
+
+		preBuiltNode := pkgGraph.CloneNode(node)
+		preBuiltNode.State = pkggraph.StateUpToDate
+		preBuiltNode.Type = pkggraph.TypePreBuilt
+
+		parentNodes := pkgGraph.To(node.ID())
+		for parentNodes.Next() {
+			parentNode := parentNodes.Node().(*pkggraph.PkgNode)
+
+			if parentNode.Type != pkggraph.TypeGoal {
+				pkgGraph.RemoveEdge(parentNode.ID(), node.ID())
+
+				logger.Log.Debugf("Adding a 'PreBuilt' node '%s' with id %d. For '%s'", preBuiltNode.FriendlyName(), preBuiltNode.ID(), parentNode.FriendlyName())
+				err = pkgGraph.AddEdge(parentNode, preBuiltNode)
+
+				if err != nil {
+					logger.Log.Errorf("Adding edge failed for %v -> %v", parentNode, preBuiltNode)
+				}
+			}
+		}
+	}
+
+	return
+}
+
+func removeExistingGoalNode(pkgGraph *pkggraph.PkgGraph, goalNodeName string) {
+	goalNode := pkgGraph.FindGoalNode(goalNodeName)
+	if goalNode != nil {
+		logger.Log.Infof("Found goalNode: %s. Removing it.", goalNodeName)
+		pkgGraph.RemoveNode(goalNode.ID())
+	} else {
+		logger.Log.Warnf("Can't find goalNode: %s. Skipping Removal", goalNodeName)
+	}
+	return
+}
+
+func subGraphPkgsToBuild(pkgGraph *pkggraph.PkgGraph, packageVersToBuild []*pkgjson.PackageVer) (optimizedGraph *pkggraph.PkgGraph, err error) {
+	optimizedGraph = pkgGraph
+
+	pkgToBuildGoalNode := pkgGraph.FindGoalNode(allGoalNodeName)
+	if schedulerutils.CanSubGraph(pkgGraph, pkgToBuildGoalNode, true) {
+		optimizedGraph, err = pkgGraph.CreateSubGraph(pkgToBuildGoalNode)
+		if err != nil {
+			logger.Log.Panicf("Failed to create subgraph error: %s", err)
+			return
+		}
+	} else {
+		logger.Log.Warnf("Not possible to subgraph")
+		return
+	}
+
+	return
+}
+
+func main() {
+	app.Version(exe.ToolkitVersion)
+	kingpin.MustParse(app.Parse(os.Args[1:]))
+	logger.InitBestEffort(*logFile, *logLevel)
+
+	scrubbedGraph := pkggraph.NewPkgGraph()
+
+	err := pkggraph.ReadDOTGraphFile(scrubbedGraph, *inputGraphFile)
+	if err != nil {
+		logger.Log.Panicf("Failed to read graph to file, %s. Error: %s", *inputGraphFile, err)
+	}
+
+	if !*hydratedBuild {
+		err = pkggraph.WriteDOTGraphFile(scrubbedGraph, *outputGraphFile)
+		if err != nil {
+			logger.Log.Panicf("Failed to write cache graph to file, %s. Error: %s", *outputGraphFile, err)
+		}
+		return
+	}
+
+	// Generate the list of packages that need to be built.
+	// If none are requested then all packages will be built.
+	packagesNamesToBuild := exe.ParseListArgument(*pkgsToBuild)
+	packagesNamesToRebuild := exe.ParseListArgument(*pkgsToRebuild)
+
+	pkgVersToBuild, err := schedulerutils.CalculatePackagesToBuild(packagesNamesToBuild, packagesNamesToRebuild, *imageConfig, *baseDirPath)
+	if err != nil {
+		logger.Log.Panicf("Error while calculating to build. Error: %s", err)
+	}
+
+	removeExistingGoalNode(scrubbedGraph, allGoalNodeName)
+	_, err = scrubbedGraph.AddGoalNode(allGoalNodeName, pkgVersToBuild, true)
```

Comment:
I think this should be after the prebuilt nodes are swapped in. (I'm assuming we need to remove the goal node because it points to the previous nodes?)

That said, I think the replaceRunNodes... function may re-assign the edges from the existing goal node as is?

Reply:
Actually, the scheduler will ignore this goal generally. It adds its own goal node to bassed on the package build lists.